### DISCUSSION TITLE
`gix clean` - the `git clean` I never knew I wanted

### DISCUSSION THREAD

**Discussion by Byron** [GitoxideLabs/gitoxide#1308] [Show and tell]
`git clean` is a fantastic tool - it makes it possible to dissect all files in the working tree that are not tracked, and chose them for deletion. Very practical as data just keeps accumulating.

### `git clean` - how I used it

When running it on `gitoxide`, and I keep making this mistake, I am typically bumping against a fair reminder:

```
❯ git clean
fatal: clean.requireForce defaults to true and neither -i, -n, nor -f given; refusing to clean
```

Of course, how could I forget, again, to at least add `-n` to explicitly enable `—dry-run` mode.

```
❯ git clean -n
```

No output. Now would be a good time to study the manual or add `—help`, and most of us probably have developed muscle memory that will crank up the amount of directory entries it will print, for a first overview.

```
❯ git clean -nxd
Would remove .DS_Store
Would remove .cargo/config.toml.mine
Would remove .idea/
Would remove .vscode/
Would remove gix-attributes/fuzz/artifacts/
[..]
Would skip repository gix-dir/tests/fixtures/generated-do-not-edit/many/2067822798-unix/type-mismatch
[..]
Would remove gix-dir/tests/fixtures/generated-do-not-edit/many/2067822798-unix/nonstandard-worktree-untracked/seemingly-outside
[..] Would skip repository gix-dir/tests/fixtures/generated-do-not-edit/many/2067822798-unix/nonstandard-worktree/dir-with-dot-git
[..]
Would remove gix-dir/tests/fixtures/generated-do-not-edit/many-symlinks/1468042303-unix/symlink-to-breakout-symlink
Would remove gix-object/fuzz/artifacts/
[..]
Would remove gix-url/fuzz/artifacts/
Would skip repository gix/tests/fixtures/generated-do-not-edit/make_basic_repo/192414529-unix
[..]
Would skip repository gix/tests/fixtures/generated-do-not-edit/make_rebase_i_repo/1880667692-unix
Would remove target/
[..]
Would remove tests/fixtures/repos/rust.git
Would remove tests/fixtures/repos/linux.git
Would remove untracked-with-nested-bare/
```

In the `gitoxide` repository, it prints hundreds of lines, most of them are messages like “Would skip repository…” that make its output incredibly noisy and hard to understand. One can force their removal by passing another `-f`, so `-ff`, but unfortunately there is no way to *not* display their presence in dry-run mode, with `-nn` maybe. This alone makes me dread its usage. That automatic repository search in what constitutes ignored directories also isn’t free, as it now has to traverse thousands of additional directories causing visible delay.

But there is more, as some of this output is probably typical for many of its runs in other repositories as well.

First off, it’s unclear *why* it wants to remove the entries it shows. Of course, these candidates are not tracked, but are they ignored, or are they untracked? Typically, untracked files are less likely to be removable, and ideally these are easy to discern.

Then, it says that it `Would remove .idea/` and that it `Would remove .vscode/`, which are ignored directories in every repository. It’s less obvious that patterns in `.gitignore` files cause files and directories to be ignored and *expendable*. `.idea/` and `.vscode/` definitely should not be tracked, but unfortunately that also puts them on the default kill list of `git clean`. Of course, `git clean` isn’t at fault here, but I mention it because it’s a usability issue everyone has to content with, and… mistakes happen all the time.

Despite trying hard to skip repositories, it is happy to `remove tests/fixtures/repos/linux.git` for example, which is an ignored fixture I use for local manual testing from time to time (`tests/fixtures/repos/` is in `.gitignore`). And it is a bare repository, which `git clean` doesn’t detect which seems surprising given all the prior chatter about ignored repositories that it refuses to remove.

#### Summary

Here are the shortcomings for using `git clean` for cleaning the `gitoxide` working directory, roughly in order of severity.

* the display of non-bare repositories in ignored directories is too noisy, and dominates the output. There is no way to turn that off.
* it wants to remove ignored files that are important to me (i.e. `.idea/` and `.vscode/`)
* it wants to remove ignored bare repositories that are fully ignored (i.e. `tests/fixtures/repos/linux.git`)
* it doesn’t visually communicate what’s ignored and what’s untracked - this makes a difference as untracked files are typically not expendable, or should be treated with greater care
* by default it refuses any work, and even with `-n` it’s easy to see nothing on screen which might keep new users puzzled

### Enter `gix clean`

With all my grievances	with `git clean`, right after finishing `gix_dir::walk()` it seemed like an obvious challenge to try the ‘walk’ implementation on. And while at it, I thought I could make a few improvements on the way. First of all, the initial implementation of `gix clean` took about two days, and another day for polishing after more shortcomings in `gix_dir::walk()` were unearthed, tested for and fixed. This probably speaks for the overall power held in the dirwalk implementation, which makes it easy to set tools on top. 

(*It also seems fair to note that it took many more days on top of that to fix many more issues that I noticed after using `gix clean` more, so it's not like the dirwalk was perfectly working in every which way from the beginning*.)

```
❯ gix clean
Nothing to clean (Skipped 1 directory - show with -d; Skipped 4 expendable entries - show with -x; Skipped 24 precious entries - show with -p)
```

It starts out with `-n` implied, and instead of printing nothing, it actively suggests which flags to add in order to show additional entries. Let’s go with `-dx` to get closer to the typical `git clean -nxd` invocation.

```
❯ gix clean -dx
WOULD remove .DS_Store (🗑️)
WOULD remove gix-dir/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove gix/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove target/ (🗑️)
WOULD remove untracked-with-nested-bare/

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip (Skipped 24 precious entries - show with -p)
WARNING: would remove repositories hidden inside untracked directories - use --find-untracked-repositories to find
```
*(This is the entire output, nothing was truncated from it)*

This looks a lot cleaner. Thanks to the use of (debatable) emoji, it makes clear which files are expandable and which are untracked. It does not show or look for repositories contained in ignored *or* untracked directories, but loudly tells the user about it while offering a remedy in the form of `--skip-hidden-repositories` and`--find untracked-repositories`.

And now, it’s definitely time to wonder why it didn’t list any o the `tests/fixtures/<bare-repository` entries, or how it could exclude the editor configuration.
Let’s take another look with `-p`, showing *precious* files.

```
❯ gix clean -dxp
WOULD remove .DS_Store (🗑️)
WOULD remove .cargo/config.toml.mine (💲)
WOULD remove .idea/ (💲)
WOULD remove .vscode/ (💲)
WOULD remove gix-attributes/fuzz/artifacts/ (💲)
WOULD remove gix-attributes/fuzz/corpus/ (💲)
WOULD remove gix-commitgraph/fuzz/Cargo.lock (💲)
[..]
WOULD remove gix-config/fuzz/corpus/ (💲)
WOULD remove gix-dir/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove gix-object/fuzz/artifacts/ (💲)
[..]
WOULD remove gix-url/fuzz/artifacts/ (💲)
WOULD remove gix/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove target/ (🗑️)
WOULD remove tests/fixtures/commit-graphs/ (💲)
WOULD remove tests/fixtures/repos/ (💲)
WOULD remove untracked-with-nested-bare/

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip
WARNING: would remove repositories hidden inside untracked directories - use --find-untracked-repositories to find
```

Using the experimental feature that is ready for first implementation in Git as well, I was able to declare a subset of ignored files as *precious*. With that knowledge, there is no risk of accidentally nuking important ignored files anymore. For instance, to make `.vscode/` precious, edit the `.gitignore` file that contains the line to ignore it, to look like this:

```gitignore
.vscode/
$.vscode/
```

Now `gitoxide` will see these entries as *precious*, while Git (or older Git once it supports it) will meaningfully see only `.vscode/` and treat it as ignored and expendable like before.

Going back, let’s be sure that we don’t remove repositories in untracked directories.

```
❯ gix clean -dx --find-untracked-repositories=all
WOULD remove .DS_Store (🗑️)
WOULD remove gix-dir/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove gix/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove target/ (🗑️)
WOULD remove untracked-with-nested-bare/.DS_Store (🗑️)
WOULD remove untracked-with-nested-bare/file
WOULD remove untracked-with-nested-bare/subdir/.DS_Store (🗑️)
WOULD remove untracked-with-nested-bare/subdir/deeper/.DS_Store (🗑️)

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip (Skipped 1 repository - show with -r; Skipped 24 precious entries - show with -p)
```

Now the warning message disappears, and instead if says that it `Skipped 1 repositories`, while avoiding to collapse the `untracked-with-nested-bare’ directory, instead showing which files it can delete without affecting the repository within.

If we wanted to remove that repository too, `-r` can be specified.

```
❯ gix clean -dx --find-untracked-repositories=all -r
[..]
WOULD remove repository untracked-with-nested-bare/subdir/deeper/direct-bare/

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip (Skipped 24 precious entries - show with -p)
```

Typically I am most interested in removing generated files, so let’s use a *pathspec* to narrow the list we see.

```
❯ gix clean -xd  '*/generated-do-not-edit/'
WOULD remove gix/tests/fixtures/generated-do-not-edit/ (🗑️)

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip (try to adjust your pathspec to reveal some of the 108 pruned entries - show with --debug)
```

What’s going on though? I would have expected two directories, so the line `WOULD remove gix-dir/tests/fixtures/generated-do-not-edit/ (🗑️)` is unexpectedly missing. The message fortunately tells us about 108 pruned entries that the pathspec doesn’t entail, and that can be shown using `—debug`.

When doing that one will see that the *pathspec* actually affects the directory walk, and wildcards in *pathspecs* will cause the search to continue into otherwise ignored directories, fundamentally changing the set of returned files.

Of course, we could adjust the *pathspec* to `*/generated-do-not-edit/*` to also list the pruned entries, but then we are back to a noisy display that originally we tried to avoid. Now also `gix clean` will detect the ignored non-bare directories, and thus fail to collapse the folder like before.

In order to get the results that was originally intended, there has to be a way to apply the *pathspec* to *displayed entries only*. And with `-m` we do just that.

```
❯ gix clean -xd  '*/generated-do-not-edit/' target -m
WOULD remove gix-dir/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove gix/tests/fixtures/generated-do-not-edit/ (🗑️)
WOULD remove target/ (🗑️)

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip (try to adjust your pathspec to reveal some of the 27 pruned entries - show with --debug)
```

I also added `target` to the *pathspec* for good measure, and believe that this is now the set to delete. Adding `-e` for `—execute` will do the trick.

```
❯ gix clean -xd  '*/generated-do-not-edit/' target -m -e
removing gix-dir/tests/fixtures/generated-do-not-edit/ (🗑️)
removing gix/tests/fixtures/generated-do-not-edit/ (🗑️)
removing target/ (🗑️)
```

### Conclusion

`gix clean` is built on the new `gix-dir` crate which provides a configurable Git-style directory walk. Its 1100SLOC are supported by more than 4000SLOC of tests to pin-point every bit of logic, and yet, there is probably still a test or two missing as real-world usage might point out later.

While working on it, I gained an even greater respect for the accomplishment of `git clean` as is, given how much it gets right, and how many edge-cases it handles gracefully. This was particularly noticeable when I simply had to *keep adding* features to merely be en-par with what Git can already do.

At some point I did also realise how many shortcuts Git takes in the name of optimisations, and how this actually hurts the general-purposes-ness of the library that `gix-dir` is going to be. Step by step `gitoxide`’s model evolved to become more general. Ultimately this helped `gix clean` to gain access to much richer and consistent information about the directory, and to keep its own logic simple. And with that simplicity as baseline, it was possible to add complexity for a better user experience.

Hopefully, this is also just the beginning :).

### Bonus - CWD handling

`git clean` will assure that it won’t delete the directory you are currently in, showing incredible attention to detail and love for the command-line.

<details><summary>Expand for CWD details</summary>

```
❯ cd target/tmp
❯ git clean -nxd ../
Would refuse to remove current working directory
Would remove ../.rustc_info.json
Would remove ../release
Would remove ../debug
```

`gix clean` inherited this, while keeping up its reporting detail.

```
❯ gix clean -xd ../
WOULD remove ../.rustc_info.json (🗑️)
WOULD remove ../debug/ (🗑️)
WOULD remove ../release/ (🗑️)
Would refuse to remove empty current working directory ./ (🗑️)

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip
```

When entering a populated directory, the result is a little bit different.

```
❯ cd ../debug
❯ git clean -nxd ../
Would refuse to remove current working directory
Would remove ../debug/.fingerprint
Would remove ../debug/incremental
Would remove ../debug/gix.d
Would remove ../debug/gix
Would remove ../debug/.cargo-lock
Would remove ../debug/examples
Would remove ../debug/deps
Would remove ../debug/build
Would remove ../.rustc_info.json
Would remove ../release
Would remove ../tmp
```

Now Git will expand the entries in our current working directory, `target/debug`, but displays it in an unusual form, I would have expected the `./.fingerprint` notation for example, not `../debug/.fingerprint`.

Compare this to `gix clean`:

```
❯ gix clean -xd ../
WOULD remove ../.rustc_info.json (🗑️)
WOULD remove .cargo-lock (🗑️)
WOULD remove .fingerprint/ (🗑️)
WOULD remove build/ (🗑️)
WOULD remove deps/ (🗑️)
WOULD remove examples/ (🗑️)
WOULD remove gix (🗑️)
WOULD remove gix.d (🗑️)
WOULD remove empty incremental/ (🗑️)
WOULD remove ../release/ (🗑️)
WOULD remove empty ../tmp/ (🗑️)

WARNING: would remove repositories hidden inside ignored directories - use --skip-hidden-repositories to skip
```

It also prevented the collapse of the current working directory, but displays paths as we would expect.

</details>

### Bonus - Performance

This is clearly not an apples to apples comparison as the logic of `gix clean` is different in one major way: it won’t search for hidden repositories by default. But with that enabled, how will it fare?

<details><summary>Expand for Benchmarks</summary>

#### `gitoxide` repository@aa7c1908b82e3e23859a4c663faa40ec54611919 - default settings

In order to get similar state, run `cargo test -p gix dir && cargo test -p gix`.
In a way, the out-of-the-box performance with default settings matters to me, even though it’s not a fair comparison. That one follows.

```
❯ hyperfine -w1 -N 'git clean -nxd' 'gix clean -xd'
Benchmark 1: git clean -nxd
  Time (mean ± σ):     233.6 ms ±   5.1 ms    [User: 46.6 ms, System: 180.7 ms]
  Range (min … max):   228.4 ms … 245.9 ms    12 runs

Benchmark 2: gix clean -xd
  Time (mean ± σ):      17.1 ms ±   0.9 ms    [User: 5.3 ms, System: 8.8 ms]
  Range (min … max):    15.7 ms …  23.3 ms    172 runs

Summary
  gix clean -xd ran
   13.63 ± 0.78 times faster than git clean -nxd
```

`gix` is a lot faster here because it does way less work out of the box.

#### `gitoxide` repository@aa7c1908b82e3e23859a4c663faa40ec54611919 - apples-to-apples

```
gitoxide ( status) [$?]
❯ hyperfine -w1 -N 'git clean -nxd' 'gix clean -xd --skip-hidden-repositories=non-bare'
Benchmark 1: git clean -nxd
  Time (mean ± σ):     232.6 ms ±   4.2 ms    [User: 46.7 ms, System: 181.1 ms]
  Range (min … max):   228.4 ms … 243.8 ms    13 runs

Benchmark 2: gix clean -xd --skip-hidden-repositories=non-bare
  Time (mean ± σ):     153.9 ms ±   2.4 ms    [User: 53.1 ms, System: 96.4 ms]
  Range (min … max):   151.4 ms … 161.1 ms    19 runs

Summary
  gix clean -xd --skip-hidden-repositories=non-bare ran
    1.51 ± 0.04 times faster than git clean -nxd
```

Here it reads 3065 directories and sees 36799 filesystem entries. Git also sees 3065 directories along with 37516 entries, but takes significantly longer for that, for some reason. Maybe it’s related to the way it detects non-bare repositories?

#### `linux` @ffc2532

This is a clean working tree.

```
❯ hyperfine -w1 -N 'git clean -nxd' 'gix clean -xd --skip-hidden-repositories=non-bare'
Benchmark 1: git clean -nxd
  Time (mean ± σ):     137.8 ms ±   4.3 ms    [User: 39.1 ms, System: 104.2 ms]
  Range (min … max):   133.7 ms … 151.5 ms    22 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet system without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 2: gix clean -xd --skip-hidden-repositories=non-bare
  Time (mean ± σ):     173.8 ms ±   5.5 ms    [User: 69.9 ms, System: 103.9 ms]
  Range (min … max):   166.5 ms … 186.7 ms    17 runs

Summary
  git clean -nxd ran
    1.26 ± 0.06 times faster than gix clean -xd --skip-hidden-repositories=non-bare
```

On bigger repositories, Git seems to gain the advantage. It traverses 5317 directories with 87071 entries. `gix` sees the same amount of directories with 81755 entries, but takes longer overall, particularly in user land.

When disabling a few bottlenecks of the `gitoxide` implementation, `ignoreCase` and `precomposeUnicode`, we see this:

```
❯ hyperfine -w1 -N 'git -c core.ignoreCase=0 -c core.precomposeUnicode=0 clean -nxd' 'gix -c core.ignoreCase=0 -c core.precomposeUnicode=0 clean -xd --skip-hidden-repositories=non-bare'
Benchmark 1: git -c core.ignoreCase=0 -c core.precomposeUnicode=0 clean -nxd
  Time (mean ± σ):     137.9 ms ±   2.3 ms    [User: 35.6 ms, System: 102.1 ms]
  Range (min … max):   133.5 ms … 142.4 ms    22 runs

Benchmark 2: gix -c core.ignoreCase=0 -c core.precomposeUnicode=0 clean -xd --skip-hidden-repositories=non-bare
  Time (mean ± σ):     165.4 ms ±   4.2 ms    [User: 62.7 ms, System: 102.5 ms]
  Range (min … max):   159.6 ms … 175.7 ms    18 runs

Summary
  git -c core.ignoreCase=0 -c core.precomposeUnicode=0 clean -nxd ran
    1.20 ± 0.04 times faster than gix -c core.ignoreCase=0 -c core.precomposeUnicode=0 clean -xd --skip-hidden-repositories=non-bare
```

That’s just a little better, this round clearly goes to Git :).

#### WebKit @ 886077e077a496a6e398df52a4b7915d8cd68f76

This is a plain checkout without changes to the working tree. However, I have turned off `core.ignoreCase` and `core.precomposeUnicode` in the repository configuration as it makes a difference for both applications (but more so for `gix`).

```
❯ hyperfine -w1 -N 'git clean -nxd' 'gix clean -xd --skip-hidden-repositories=non-bare'
Benchmark 1: git clean -nxd
  Time (mean ± σ):     753.0 ms ±  12.5 ms    [User: 233.0 ms, System: 507.1 ms]
  Range (min … max):   741.3 ms … 784.1 ms    10 runs

Benchmark 2: gix clean -xd --skip-hidden-repositories=non-bare
  Time (mean ± σ):     801.2 ms ±   5.1 ms    [User: 294.6 ms, System: 496.4 ms]
  Range (min … max):   795.3 ms … 810.7 ms    10 runs

Summary
  git clean -nxd ran
    1.06 ± 0.02 times faster than gix clean -xd --skip-hidden-repositories=non-bare
```

Here Gix is just a little bit slower, but that gets much worse once the aforementioned flags are turned on.

```
❯ hyperfine -w1 -N 'git -c core.ignoreCase=1 -c core.precomposeUnicode=1 clean -nxd' 'gix -c core.ignoreCase=1 -c core.precomposeUnicode=1 clean -xd --skip-hidden-repositories=non-bare'
Benchmark 1: git -c core.ignoreCase=1 -c core.precomposeUnicode=1 clean -nxd
  Time (mean ± σ):     722.0 ms ±   8.3 ms    [User: 231.7 ms, System: 510.9 ms]
  Range (min … max):   712.6 ms … 741.3 ms    10 runs

Benchmark 2: gix -c core.ignoreCase=1 -c core.precomposeUnicode=1 clean -xd --skip-hidden-repositories=non-bare
  Time (mean ± σ):      1.014 s ±  0.005 s    [User: 0.504 s, System: 0.498 s]
  Range (min … max):    1.008 s …  1.023 s    10 runs

Summary
  git -c core.ignoreCase=1 -c core.precomposeUnicode=1 clean -nxd ran
    1.40 ± 0.02 times faster than gix -c core.ignoreCase=1 -c core.precomposeUnicode=1 clean -xd --skip-hidden-repositories=non-bare
```

Now Git is 1.4 times faster even, and it’s good to know that `gix` still has some untapped potential for performance here. Git is highly optimised in this area, and even uses multi-threading in a place where `gix` doesn’t (yet).

#### Git @ 3c2a3fdc388747b9eaf4a4a4f2035c1c9ddb26d0

The working tree is the one after a build, with plenty of build-artifacts littered everywhere.

```
❯ hyperfine -w1 -N 'git clean -nxd' 'gix clean -xd --skip-hidden-repositories=non-bare'
Benchmark 1: git clean -nxd
  Time (mean ± σ):      21.8 ms ±   1.0 ms    [User: 7.3 ms, System: 11.9 ms]
  Range (min … max):    20.2 ms …  28.0 ms    131 runs

Benchmark 2: gix clean -xd --skip-hidden-repositories=non-bare
  Time (mean ± σ):      20.1 ms ±   0.9 ms    [User: 8.1 ms, System: 8.7 ms]
  Range (min … max):    18.6 ms …  23.6 ms    146 runs

Summary
  gix clean -xd --skip-hidden-repositories=non-bare ran
    1.09 ± 0.07 times faster than git clean -nxd
```

Despite taking more CPU time in user land, `gix` takes much less time of the system for some reason, to be a little faster over all.

#### Verdict

It seems that for small to medium repositories, `gix clean` will be a little faster, while being up to 50% slower on very large repositories.

</details>

#### Precious Files?

In my own words, precious files are ‘the missing class’ of ignored files. Currently, all ignored files are also expendable. But precious files are ignored and… *not* expendable which makes them closer to untracked files.

They can be specified by prefixing a pattern in `.gitignore` files with `$`. Git (or older Git when it supports that) sees these entries verbatim, but `gix` will know it’s precious. If it sees such a line before one that marks it as expendable, it will have the desired effect.

To retrofit a `.gitignore` file that works normally for all, let the precious pattern *follow* the expendable one.

```.gitignore
normal
# this is picked up by those who support precious files, or otherwise will lead to ignored files actually prefixed with $ which usually doesn’t matter.
$normal
```

The technical document for all the details is [available on the mailing list](https://lore.kernel.org/git/pull.1627.git.1703643931314.gitgitgadget@gmail.com/) or [on GitHub](https://github.com/newren/git/blob/0e6e3a6fd3c86be75963392c1c99f1c8ac704fa8/Documentation/technical/precious-files.txt), authored by Elijah Newren.

#### Should I use `gix clean`?

I think using it makes a major difference once *precious files* are leveraged. Because now everything will look so much cleaner that you wouldn’t want to miss it anymore. Without precious files it makes much less of a difference, but is still easier to use overall particularly with the `-m` flag.

You can download the latest release here to give it a try: https://github.com/Byron/gitoxide/releases/tag/v0.34.0

**Comment by NobodyXu** [comment]
It's a surpirse to me that ignored files will be removed by git-clean.

I think it might be reasonable to put them into a trash bin instead of removing immediately?
Though that will be platform dependent.

**Reply by Byron** [reply]
Yeah, I don't know what's causing this, but already validated that it has nothing to do with my precious-file setup.

> I think it might be reasonable to put them into a trash bin instead of removing immediately?
> Though that will be platform dependent.


There is a crate that I took into my custody called `trash-rs` which does exactly that. I would leave its usage to other applications though, it's easy enough to do now :).

**Reply by TheFibonacciEffect** [reply]
Oh that might be a bad idea, because if I have a code that is generating a large output, ill certainly ignore the output data file. I would not want that much data to end up in trash.
